.github/workflows/main.yml

```diff
@@ -0,0 +1,89 @@
+    outputs:
+      STABLE_P4C_COMMIT: ${{ steps.solving_envs.outputs.STABLE_P4C_COMMIT }}
+      STABLE_20210108_P4C_COMMIT: ${{ steps.solving_envs.outputs.STABLE_20210108_P4C_COMMIT }}
+      LATEST_P4C_COMMIT: ${{ steps.solving_envs.outputs.LATEST_P4C_COMMIT }}
+      MAKEFLAGS: ${{ steps.solving_envs.outputs.MAKEFLAGS }}
+      DOCKER_REPO: ${{ steps.solving_envs.outputs.DOCKER_REPO }}
+        
+        
+  build_matrix:
+    needs: solve_env
+    runs-on: ubuntu-20.04 
+    strategy:
+      matrix:
+          include:
+          - TAGNAME: latest
+            P4C_COMMIT: ${{ needs.solve_env.outputs.LATEST_P4C_COMMIT }}
+            PROTOBUF_VERSION: 3.6.1
+          - TAGNAME: stable
+            P4C_COMMIT: ${{ needs.solve_env.outputs.STABLE_P4C_COMMIT }}
+            PROTOBUF_VERSION: 3.2.0
+          - TAGNAME: stable-20210108
+            P4C_COMMIT: ${{ needs.solve_env.outputs.STABLE_20210108_P4C_COMMIT }}
+            PROTOBUF_VERSION: 3.6.1
+            
+    steps:
+      - name: Checkout this repository
+        uses: actions/checkout@v2
+        
+      - name: Before Install
+        shell: bash
+        id: before_install
+        run: |
+          echo "::set-output name=LOCAL_RUNTIME_IMAGE::p4c-${{ matrix.TAGNAME }}"
+          echo "::set-output name=DOCKER_RUNTIME_IMAGE::${{ needs.solve_env.outputs.DOCKER_REPO }}:${{ matrix.TAGNAME }}"
+          echo '::set-output name=BUILD_ARGS::--build-arg PROTOBUF_VERSION=${{ matrix.PROTOBUF_VERSION }} --build-arg P4C_COMMIT=${{ matrix.P4C_COMMIT }} --build-arg MAKEFLAGS=${{ needs.solve_env.outputs.MAKEFLAGS }}'
+      
+      - name: Docker Pull Image
+        id: docker_pull
+        run: docker pull ${{ steps.before_install.outputs.DOCKER_RUNTIME_IMAGE }}
+        
+      - name: Run scripts
+        id: run_scripts
+        run: |
+            # Build images. Without eval, build args are not escaped properly.
+            eval "docker build --target runtime --cache-from ${{ steps.before_install.outputs.DOCKER_RUNTIME_IMAGE }} -t ${{ steps.before_install.outputs.LOCAL_RUNTIME_IMAGE }} ${{ steps.before_install.outputs.BUILD_ARGS }} . "
+            
+      - name: Test docker build
+        run: |
+          # Test that all shared objects are present.
+          docker run --rm ${{ steps.before_install.outputs.LOCAL_RUNTIME_IMAGE }} p4c-bm2-ss --help
+          docker run --rm ${{ steps.before_install.outputs.LOCAL_RUNTIME_IMAGE }} p4c-graphs --help
+
+      - name: Change file permissions
+        run: chmod +x docker_push
+        
+      - name: Run docker_push
+        run: ./docker_push "$DOCKER_PASSWORD" "$DOCKER_USERNAME" "$DOCKER_RUNTIME_IMAGE" "$LOCAL_RUNTIME_IMAGE"
+        env:
+          DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+          DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+          DOCKER_RUNTIME_IMAGE: ${{ steps.before_install.outputs.DOCKER_RUNTIME_IMAGE }}
+          LOCAL_RUNTIME_IMAGE : ${{ steps.before_install.outputs.DOCKER_RUNTIME_IMAGE }}
```

Comment:
As we discussed, we can most probably directly use the GH action do do build and pull

Reply:
Yes. I'll try to use https://github.com/marketplace/actions/build-and-push-docker-images